modules/client-common/src/main/java/org/apache/ignite/internal/client/proto/ClientComputeJobPacker.java

```diff
@@ -0,0 +1,203 @@
+import java.util.UUID;
+import org.apache.ignite.internal.binarytuple.inlineschema.TupleWithSchemaMarshalling;
+import org.apache.ignite.marshalling.Marshaller;
+import org.apache.ignite.sql.ColumnType;
+import org.apache.ignite.table.Tuple;
+import org.jetbrains.annotations.Nullable;
+
+/** Packs job arguments and results. */
+public class ClientComputeJobPacker {
+    private final ClientMessagePacker packer;
+
+    /**
+     * Constructor.
+     *
+     * @param packer that should be closed after usage outside of this class.
+     */
+    public ClientComputeJobPacker(ClientMessagePacker packer) {
+        this.packer = packer;
+    }
+
+    /**
+     * Packs compute job argument. If the marshaller is provided, it will be used to marshal the argument. If the marshaller is not provided
+     * and the argument is a native column type or a tuple, it will be packed accordingly.
+     *
+     * @param arg Argument.
+     * @param marshaller Marshaller.
+     * @param <T> Argument type.
+     */
+    public <T> void packJobArgument(@Nullable T arg, @Nullable Marshaller<T, byte[]> marshaller) {
+        pack(arg, marshaller);
+    }
+
+    /**
+     * Packs compute job result. If the marshaller is provided, it will be used to marshal the result. If the marshaller is not provided and
+     * the result is a native column type or a tuple, it will be packed accordingly.
+     *
+     * @param res Result.
+     * @param marshaller Marshaller.
+     * @param <T> Result type.
+     */
+    public <T> void packJobResult(@Nullable T res, @Nullable Marshaller<T, byte[]> marshaller) {
+        pack(res, marshaller);
+    }
+
+    /** Packs object in the format: | typeId | value |. */
+    private <T> void pack(@Nullable T obj, @Nullable Marshaller<T, byte[]> marshaller) {
+        if (marshaller != null) {
+            packer.packInt(ComputeJobType.MARSHALLED_OBJECT_ID);
+            byte[] marshalled = marshaller.marshal(obj);
+
+            if (marshalled == null) {
+                packer.packNil();
+                return;
+            }
+
+            packer.packByteBuffer(ByteBuffer.wrap(marshalled));
+            return;
+        }
+
+        if (obj instanceof Tuple) {
+            byte[] marshalledTuple = TupleWithSchemaMarshalling.marshal((Tuple) obj);
+
+            packer.packInt(ComputeJobType.MARSHALLED_TUPLE_ID);
+
+            if (marshalledTuple == null) {
+                packer.packNil();
+                return;
+            }
+
+            packer.packByteBuffer(ByteBuffer.wrap(marshalledTuple));
+            return;
+        }
+
+        if (obj == null) {
+            packer.packInt(ColumnType.NULL.id());
+            packer.packNil();
+            return;
+        }
+
+        var type = getColumnTypeFrom(obj);
+        assert type != null;
+
+        packer.packInt(type.id());
+
+        var writer = writerForType(type);
+        writer.write(obj, packer);
+    }
+
+    private static Writer writerForType(ColumnType type) {
```

Comment:
Allocating a `Writer` here seems unnecessary. We can refactor the method to perform the actual write within the `switch`:

```java
       switch (type) {
            case BOOLEAN:
                packer.packBoolean(((Boolean) obj));
                return;
 ...
```